File: app/src/main/java/me/liuyun/bjutlgn/widget/GraphCard.java
package me.liuyun.bjutlgn.widget;

import android.support.v7.widget.CardView;
import android.util.SparseIntArray;

import com.android.settingslib.graph.UsageView;

import java.text.DateFormat;
import java.util.Calendar;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import me.liuyun.bjutlgn.R;
import me.liuyun.bjutlgn.WiFiApplication;
import me.liuyun.bjutlgn.entity.Flow;
import me.liuyun.bjutlgn.util.StatUtils;

public class GraphCard {
    @BindView(R.id.chart) UsageView chart;
    private WiFiApplication context;
    private List<Flow> flowList;

    public GraphCard(CardView cardView, WiFiApplication context) {
        this.context = context;
        ButterKnife.bind(this, cardView);
    }

    public void show() {
        int month = Calendar.getInstance().get(Calendar.MONTH);
        if (context.getPrefs().getInt("current_month", -1) != month) {
            context.getPrefs().edit().putInt("current_month", month).apply();
            context.getFlowManager().clearFlow();
        }

        flowList = context.getFlowManager().getAllFlow();
        chart.clearPaths();
        chart.configureGraph(60 * 24 * getEndOfCurrentMonth().get(Calendar.DATE),  StatUtils.getPack(context) * 1024, true, true);
        calcPoints();

        DateFormat dateFormat = DateFormat.getDateInstance(DateFormat.SHORT);
        chart.setBottomLabels(new String[]{dateFormat.format(getStartOfCurrentMonth().getTime()),
                dateFormat.format(getEndOfCurrentMonth().getTime())});
    }

    private void calcPoints() {
        int startOfMonth = (int) (getStartOfCurrentMonth().getTimeInMillis() / 1000L) / 60;
        SparseIntArray points = new SparseIntArray();
        points.put(0, 0);
        if (flowList != null) {
            for (Flow flow : flowList) {
                points.put((int) (flow.getTimestamp() / 60 - startOfMonth), flow.getFlow() / 1024);
            }
            chart.addPath(points);
        }
    }

    private Calendar getStartOfCurrentMonth() {
        Calendar cal = Calendar.getInstance();
        cal.set(Calendar.HOUR_OF_DAY, 0);
        cal.clear(Calendar.MINUTE);
        cal.clear(Calendar.SECOND);
        cal.clear(Calendar.MILLISECOND);
        cal.set(Calendar.DAY_OF_MONTH, 1);
        return cal;
    }

    private Calendar getEndOfCurrentMonth() {
        Calendar cal = Calendar.getInstance();
        cal.set(Calendar.DATE, cal.getActualMaximum(Calendar.DATE));
        return cal;
    }
}

File: app/src/main/kotlin/me/liuyun/bjutlgn/ui/GraphCard.kt
package me.liuyun.bjutlgn.ui

import android.util.SparseIntArray
import android.widget.LinearLayout
import androidx.content.edit
import com.android.settings.graph.UsageView
import kotlinx.android.synthetic.main.graph_card.view.*
import me.liuyun.bjutlgn.App
import me.liuyun.bjutlgn.entity.Flow
import me.liuyun.bjutlgn.util.StatsUtils
import java.text.DateFormat
import java.util.*

class GraphCard(cardView: LinearLayout, private val app: App) {
    val chart: UsageView = cardView.chart

    fun show() {
        val month = Calendar.getInstance().get(Calendar.MONTH)
        if (app.prefs.getInt("current_month", -1) != month) {
            app.prefs.edit { putInt("current_month", month) }
            app.appDatabase.flowDao().deleteAll()
        }

        val flowList = app.appDatabase.flowDao().all()
        chart.clearPaths()
        chart.configureGraph(60 * 24 * endOfCurrentMonth.get(Calendar.DATE), StatsUtils.getPack(app) * 1024, true, true)
        calcPoints(flowList)

        val dateFormat = DateFormat.getDateInstance(DateFormat.SHORT)
        chart.setBottomLabels(arrayOf(dateFormat.format(startOfCurrentMonth.time), dateFormat.format(endOfCurrentMonth.time)))
    }

    private fun calcPoints(flowList: List<Flow>) {
        val startOfMonth = (startOfCurrentMonth.timeInMillis / 1000L).toInt() / 60
        val points = SparseIntArray()
        points.put(0, 0)

        flowList.forEach { points.put((it.timestamp / 60 - startOfMonth).toInt(), it.flow / 1024) }
        chart.addPath(points)
    }

    private val startOfCurrentMonth: Calendar
        get() = Calendar.getInstance().apply {
            set(Calendar.HOUR_OF_DAY, 0)
            clear(Calendar.MINUTE)
            clear(Calendar.SECOND)
            clear(Calendar.MILLISECOND)
            set(Calendar.DAY_OF_MONTH, 1)
        }

    private val endOfCurrentMonth: Calendar
        get() = Calendar.getInstance().apply {
            set(Calendar.DATE, getActualMaximum(Calendar.DATE))
        }
}
